etcdmain/config.go

```diff
@@ -77,59 +83,111 @@ type config struct {
+	CorsCfgFile    string `json:"cors"`
+
 	// TickMs is the number of milliseconds between heartbeat ticks.
 	// TODO: decouple tickMs and heartbeat tick (current heartbeat tick = 1).
 	// make ticks a cluster wide configuration.
-	TickMs            uint
-	ElectionMs        uint
-	quotaBackendBytes int64
+	TickMs            uint  `json:"heartbeat-interval"`
+	ElectionMs        uint  `json:"election-timeout"`
+	QuotaBackendBytes int64 `json:"quota-backend-bytes"`
 
 	// clustering
 	apurls, acurls      []url.URL
 	clusterState        *flags.StringsFlag
-	dnsCluster          string
-	dproxy              string
-	durl                string
+	DnsCluster          string `json:"discovery-srv"`
+	Dproxy              string `json:"discovery-proxy"`
+	Durl                string `json:"discovery"`
 	fallback            *flags.StringsFlag
-	initialCluster      string
-	initialClusterToken string
-	strictReconfigCheck bool
+	InitialCluster      string `json:"initial-cluster"`
+	InitialClusterToken string `json:"initial-cluster-token"`
+	StrictReconfigCheck bool   `json:"strict-reconfig-check"`
+	ApurlsCfgFile       string `json:"initial-advertise-peer-urls"`
+	AcurlsCfgFile       string `json:"advertise-client-urls"`
+	ClusterStateCfgFile string `json:"initial-cluster-state"`
+	FallbackCfgFile     string `json:"discovery-fallback"`
 
 	// proxy
 	proxy                  *flags.StringsFlag
-	proxyFailureWaitMs     uint
-	proxyRefreshIntervalMs uint
-	proxyDialTimeoutMs     uint
-	proxyWriteTimeoutMs    uint
-	proxyReadTimeoutMs     uint
+	ProxyFailureWaitMs     uint   `json:"proxy-failure-wait"`
+	ProxyRefreshIntervalMs uint   `json:"proxy-refresh-interval"`
+	ProxyDialTimeoutMs     uint   `json:"proxy-dial-timeout"`
+	ProxyWriteTimeoutMs    uint   `json:"proxy-write-timeout"`
+	ProxyReadTimeoutMs     uint   `json:"proxy-read-timeout"`
+	ProxyCfgFile           string `json:"proxy"`
 
 	// security
 	clientTLSInfo, peerTLSInfo transport.TLSInfo
-	clientAutoTLS, peerAutoTLS bool
+	ClientAutoTLS              bool           `json:"auto-tls"`
+	PeerAutoTLS                bool           `json:"peer-auto-tls"`
+	ClientSecurityCfgFile      securityConfig `json:"client-security"`
+	PeerSecurityCfgFile        securityConfig `json:"peer-security"`
 
-	// logging
-	debug        bool
-	logPkgLevels string
+	// Debug logging
+	Debug        bool   `json:"debug"`
+	LogPkgLevels string `json:"log-package-levels"`
 
-	// unsafe
-	forceNewCluster bool
+	// ForceNewCluster is unsafe
+	ForceNewCluster bool `json:"force-new-cluster"`
 
 	printVersion bool
 
 	autoCompactionRetention int
 
 	enablePprof bool
 
+	configFile string
+
 	ignored []string
 }
 
+type securityConfig struct {
+	CAFile        string `json:"ca-file"`
+	CertFile      string `json:"cert-file"`
+	KeyFile       string `json:"key-file"`
+	CertAuth      bool   `json:"client-cert-auth"`
+	TrustedCAFile string `json:"trusted-ca-file"`
+}
+
+func (cfg *config) setDefaultsIfNeeded() {
```

Comment:
probably we can remove the IfNeeded part if we setDefaults at line368 before we unmarshal the yaml?